### ISSUE TITLE
Fill does not work for xarray data for labels layer

### ISSUE THREAD

**Issue by ianhi** [napari/napari#2374]
## 🐛 Bug
Using `fill` is broken for the labels layer when backed by an xarray.

It doesn't fill and throws this error:
```
  File "/home/ian/mambaforge/envs/micro/lib/python3.8/site-packages/vispy/app/backends/_qt.py", line 443, in mousePressEvent
    self._vispy_mouse_press(
  File "/home/ian/mambaforge/envs/micro/lib/python3.8/site-packages/vispy/app/base.py", line 181, in _vispy_mouse_press
    ev = self._vispy_canvas.events.mouse_press(**kwargs)
  File "/home/ian/mambaforge/envs/micro/lib/python3.8/site-packages/vispy/util/event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "/home/ian/mambaforge/envs/micro/lib/python3.8/site-packages/vispy/util/event.py", line 473, in _invoke_callback
    _handle_exception(self.ignore_callback_errors,
  File "/home/ian/mambaforge/envs/micro/lib/python3.8/site-packages/vispy/util/event.py", line 471, in _invoke_callback
    cb(event)
  File "/home/ian/mambaforge/envs/micro/lib/python3.8/site-packages/napari/_qt/qt_viewer.py", line 682, in on_mouse_press
    mouse_press_callbacks(layer, event)
  File "/home/ian/mambaforge/envs/micro/lib/python3.8/site-packages/napari/utils/interactions.py", line 101, in mouse_press_callbacks
    next(gen)
  File "/home/ian/mambaforge/envs/micro/lib/python3.8/site-packages/napari/layers/labels/_labels_mouse_bindings.py", line 35, in draw
    layer.fill(layer.coordinates, new_label)
  File "/home/ian/mambaforge/envs/micro/lib/python3.8/site-packages/napari/layers/labels/labels.py", line 776, in fill
    matches = labels == old_label
  File "/home/ian/mambaforge/envs/micro/lib/python3.8/site-packages/xarray/core/dataarray.py", line 2991, in func
    f(self.variable, other_variable)
  File "/home/ian/mambaforge/envs/micro/lib/python3.8/site-packages/xarray/core/nputils.py", line 73, in array_eq
    return _ensure_bool_is_ndarray(self == other, self, other)
  File "/home/ian/mambaforge/envs/micro/lib/python3.8/site-packages/xarray/core/variable.py", line 2305, in func
    result = Variable(dims, new_data, attrs=attrs)
  File "/home/ian/mambaforge/envs/micro/lib/python3.8/site-packages/xarray/core/variable.py", line 342, in __init__
    self._dims = self._parse_dimensions(dims)
  File "/home/ian/mambaforge/envs/micro/lib/python3.8/site-packages/xarray/core/variable.py", line 603, in _parse_dimensions
    raise ValueError(
ValueError: dimensions () must have the same length as the number of data dimensions, ndim=2
```

## To Reproduce
```python
from skimage import data
from scipy import ndimage as ndi
import napari
import xarray as xr

with napari.gui_qt():
    blobs = xr.DataArray(data.binary_blobs(length=128, volume_fraction=0.1, n_dim=3), dims=('T','Y','X')).astype(float)
    viewer = napari.view_image(blobs.astype(float), name='blobs')
    labels = xr.DataArray(ndi.label(blobs)[0])
    label_layer = viewer.add_labels(labels)
    label_layer.mode = 'fill'
```
try to fill one of the blobs.

## Expected behavior
Blob is paintbucket filled.

## Environment
```
napari: 0.4.7.dev14+gc473058
Platform: Linux-5.8.0-7630-generic-x86_64-with-glibc2.31
System: Pop!_OS 20.04 LTS
Python: 3.9.1 | packaged by conda-forge | (default, Jan 26 2021, 01:34:10) [GCC 9.3.0]
Qt: 5.15.2
PyQt5: 5.15.2
NumPy: 1.19.5
SciPy: 1.6.0
Dask: 2021.01.1
VisPy: 0.6.6

OpenGL:
- GL version: 4.6 (Compatibility Profile) Mesa 20.2.6
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1920x1080, scale 1.0

Plugins:
- console: 0.0.3
- svg: 0.1.4
```

## Additional context
Partial xarray support for the labels layer added in https://github.com/napari/napari/pull/2191 which fixed the behavior when in `paint` mode.

**Comment by ianhi** [comment]
Maybe a better fix for https://github.com/napari/napari/issues/2190 would have been to check if `data` is an xarray in `_ImageBase` and if so then to do `self._data = data.values`

**Comment by jni** [comment]
Haha I was working on #2339 and thinking exactly this, and boom here comes your issue 😂

**Comment by sofroniewn** [comment]
As commented in #2376, I have concerns about that approach - I think at that point I'd just say we don't support xarray and please pass us the numpy array rather than silently convert for people. Maybe we can improve the fill more generally and add an xarray test?

**Comment by ianhi** [comment]
As an update this no longer break explicitly but does try to allocate a stupendously large array:
Use this script:
```python
import xarray as xr
import napari

import numpy as np
arr = xr.DataArray(np.zeros((50, 512, 512), dtype=np.uint16))
viewer = napari.Viewer()
viewer.add_labels(arr)
napari.run()

```

and then try to fill

```
---------------------------------------------------------------------------
MemoryError                               Traceback (most recent call last)
~/mambaforge/envs/micro/lib/python3.9/site-packages/vispy/app/backends/_qt.py in mousePressEvent(self=<vispy.app.backends._qt.CanvasBackendDesktop object>, ev=<PyQt5.QtGui.QMouseEvent object>)
    500         if self._vispy_canvas is None:
    501             return
--> 502         self._vispy_mouse_press(
        self._vispy_mouse_press = <bound method BaseCanvasBackend._vispy_mouse_press of <vispy.app.backends._qt.CanvasBackendDesktop object at 0x7fd4257a3280>>
        global native = undefined
        ev = <PyQt5.QtGui.QMouseEvent object at 0x7fd4257fd040>
        global pos = undefined
        global _get_event_xy = <function _get_event_xy at 0x7fd470812a60>
        global button = undefined
        global BUTTONMAP.get = <built-in method get of dict object at 0x7fd4708246c0>
        ev.button = <built-in method button of QMouseEvent object at 0x7fd4257fd040>
        global modifiers = undefined
        self._modifiers = <bound method QtBaseCanvasBackend._modifiers of <vispy.app.backends._qt.CanvasBackendDesktop object at 0x7fd4257a3280>>
    503             native=ev,
    504             pos=_get_event_xy(ev),

~/mambaforge/envs/micro/lib/python3.9/site-packages/vispy/app/base.py in _vispy_mouse_press(self=<vispy.app.backends._qt.CanvasBackendDesktop object>, **kwargs={'button': 1, 'buttons': [], 'last_event': <MouseEvent blocked=False button=None buttons=[]...urces=[] time=1638561113.3018866 type=mouse_move>, 'last_mouse_press': None, 'modifiers': (), 'native': <PyQt5.QtGui.QMouseEvent object>, 'pos': (499, 263), 'press_event': None})
    182         # default method for delivering mouse press events to the canvas
    183         kwargs.update(self._vispy_mouse_data)
--> 184         ev = self._vispy_canvas.events.mouse_press(**kwargs)
        ev = undefined
        self._vispy_canvas.events.mouse_press = <vispy.util.event.EventEmitter object at 0x7fd4257aa1c0>
        kwargs = {'native': <PyQt5.QtGui.QMouseEvent object at 0x7fd4257fd040>, 'pos': (499, 263), 'button': 1, 'modifiers': (), 'buttons': [], 'press_event': None, 'last_event': <MouseEvent blocked=False button=None buttons=[] delta=[0. 0.] handled=False is_dragging=False last_event=MouseEvent modifiers=() native=<PyQt5.QtGui.QMouseEvent object at 0x7fd4257fd040> pos=[499 263] press_event=None source=None sources=[] time=1638561113.3018866 type=mouse_move>, 'last_mouse_press': None}
    185         if self._vispy_mouse_data['press_event'] is None:
    186             self._vispy_mouse_data['press_event'] = ev

~/mambaforge/envs/micro/lib/python3.9/site-packages/vispy/util/event.py in __call__(self=<vispy.util.event.EventEmitter object>, *args=(), **kwargs={'button': 1, 'buttons': [], 'last_event': <MouseEvent blocked=False button=None buttons=[]...urces=[] time=1638561113.3018866 type=mouse_move>, 'last_mouse_press': None, 'modifiers': (), 'native': <PyQt5.QtGui.QMouseEvent object>, 'pos': (499, 263), 'press_event': None})
    451                     raise RuntimeError('EventEmitter loop detected!')
    452 
--> 453                 self._invoke_callback(cb, event)
        self._invoke_callback = <bound method EventEmitter._invoke_callback of <vispy.util.event.EventEmitter object at 0x7fd4257aa1c0>>
        cb = <bound method QtViewer.on_mouse_press of <napari._qt.qt_viewer.QtViewer object at 0x7fd470ad8790>>
        event = <MouseEvent blocked=False button=1 buttons=[] delta=[0. 0.] handled=False is_dragging=False last_event=MouseEvent modifiers=() native=<PyQt5.QtGui.QMouseEvent object at 0x7fd4257fd040> pos=[499 263] press_event=None source=None sources=[] time=1638561113.539664 type=mouse_press>
    454                 if event.blocked:
    455                     break

~/mambaforge/envs/micro/lib/python3.9/site-packages/vispy/util/event.py in _invoke_callback(self=<vispy.util.event.EventEmitter object>, cb=<bound method QtViewer.on_mouse_press of <napari._qt.qt_viewer.QtViewer object>>, event=<MouseEvent blocked=False button=1 buttons=[] de...urces=[] time=1638561113.539664 type=mouse_press>)
    469             cb(event)
    470         except Exception:
--> 471             _handle_exception(self.ignore_callback_errors,
        global _handle_exception = <function _handle_exception at 0x7fd4261b5310>
        self.ignore_callback_errors = False
        self.print_callback_errors = 'reminders'
        self = <vispy.util.event.EventEmitter object at 0x7fd4257aa1c0>
        global cb_event = undefined
        cb = <bound method QtViewer.on_mouse_press of <napari._qt.qt_viewer.QtViewer object at 0x7fd470ad8790>>
        event = <MouseEvent blocked=False button=1 buttons=[] delta=[0. 0.] handled=False is_dragging=False last_event=MouseEvent modifiers=() native=<PyQt5.QtGui.QMouseEvent object at 0x7fd4257fd040> pos=[499 263] press_event=None source=None sources=[] time=1638561113.539664 type=mouse_press>
    472                               self.print_callback_errors,
    473                               self, cb_event=(cb, event))

~/mambaforge/envs/micro/lib/python3.9/site-packages/vispy/util/event.py in _invoke_callback(self=<vispy.util.event.EventEmitter object>, cb=<bound method QtViewer.on_mouse_press of <napari._qt.qt_viewer.QtViewer object>>, event=<MouseEvent blocked=False button=1 buttons=[] de...urces=[] time=1638561113.539664 type=mouse_press>)
    467     def _invoke_callback(self, cb, event):
    468         try:
--> 469             cb(event)
        cb = <bound method QtViewer.on_mouse_press of <napari._qt.qt_viewer.QtViewer object at 0x7fd470ad8790>>
        event = <MouseEvent blocked=False button=1 buttons=[] delta=[0. 0.] handled=False is_dragging=False last_event=MouseEvent modifiers=() native=<PyQt5.QtGui.QMouseEvent object at 0x7fd4257fd040> pos=[499 263] press_event=None source=None sources=[] time=1638561113.539664 type=mouse_press>
    470         except Exception:
    471             _handle_exception(self.ignore_callback_errors,

~/mambaforge/envs/micro/lib/python3.9/site-packages/napari/_qt/qt_viewer.py in on_mouse_press(self=<napari._qt.qt_viewer.QtViewer object>, event=<MouseEvent blocked=False button=1 buttons=[] de...urces=[] time=1638561113.539664 type=mouse_press>)
    894             The vispy event that triggered this method.
    895         """
--> 896         self._process_mouse_event(mouse_press_callbacks, event)
        self._process_mouse_event = <bound method QtViewer._process_mouse_event of <napari._qt.qt_viewer.QtViewer object at 0x7fd470ad8790>>
        global mouse_press_callbacks = <function mouse_press_callbacks at 0x7fd471583d30>
        event = <MouseEvent blocked=False button=1 buttons=[] delta=[0. 0.] handled=False is_dragging=False last_event=MouseEvent modifiers=() native=<PyQt5.QtGui.QMouseEvent object at 0x7fd4257fd040> pos=[499 263] press_event=None source=None sources=[] time=1638561113.539664 type=mouse_press>
    897 
    898     def on_mouse_move(self, event):

~/mambaforge/envs/micro/lib/python3.9/site-packages/napari/_qt/qt_viewer.py in _process_mouse_event(self=<napari._qt.qt_viewer.QtViewer object>, mouse_callbacks=<function mouse_press_callbacks>, event=<ReadOnlyWrapper at 0x7fd4249c5bc0 for MouseEvent>)
    853         layer = self.viewer.layers.selection.active
    854         if layer is not None:
--> 855             mouse_callbacks(layer, event)
        mouse_callbacks = <function mouse_press_callbacks at 0x7fd471583d30>
        layer = <Labels layer 'arr' at 0x7fd4249aa1f0>
        event = <ReadOnlyWrapper at 0x7fd4249c5bc0 for MouseEvent at 0x7fd4249b7850>
    856 
    857     def on_mouse_wheel(self, event):

~/mambaforge/envs/micro/lib/python3.9/site-packages/napari/utils/interactions.py in mouse_press_callbacks(obj=<Labels layer 'arr'>, event=<ReadOnlyWrapper at 0x7fd4249c5bc0 for MouseEvent>)
    146         if inspect.isgenerator(gen):
    147             try:
--> 148                 next(gen)
        global next = undefined
        gen = <generator object draw at 0x7fd424a0d200>
    149                 # now store iterated genenerator
    150                 obj._mouse_drag_gen[mouse_drag_func] = gen

~/mambaforge/envs/micro/lib/python3.9/site-packages/napari/layers/labels/_labels_mouse_bindings.py in draw(layer=<Labels layer 'arr'>, event=<ReadOnlyWrapper at 0x7fd4249c5bc0 for MouseEvent>)
     40             layer.paint(coordinates, new_label)
     41         elif layer._mode == Mode.FILL:
---> 42             layer.fill(coordinates, new_label)
        layer.fill = <bound method Labels.fill of <Labels layer 'arr' at 0x7fd4249aa1f0>>
        coordinates = (0.0, 179.95039820946852, 71.67612463520015)
        new_label = 1
     43     else:  # still add an item to undo queue
     44         # when dragging, if we start a drag outside the layer, we will

~/mambaforge/envs/micro/lib/python3.9/site-packages/napari/layers/labels/labels.py in fill(self=<Labels layer 'arr'>, coord=(0.0, 179.95039820946852, 71.67612463520015), new_label=1, refresh=True)
   1116             (
   1117                 match_indices,
-> 1118                 np.array(self.data[match_indices], copy=True),
        global np.array = <built-in function array>
        self.data = <xarray.DataArray (dim_0: 50, dim_1: 512, dim_2: 512)>
<class 'numpy.ndarray'> (50, 512, 512) uint16
Dimensions without coordinates: dim_0, dim_1, dim_2
        match_indices = (<class 'numpy.ndarray'> (262144,) int64, <class 'numpy.ndarray'> (262144,) int64, <class 'numpy.ndarray'> (262144,) int64)
        global copy = undefined
   1119                 new_label,
   1120             )

~/mambaforge/envs/micro/lib/python3.9/site-packages/xarray/core/dataarray.py in __getitem__(self=<xarray.DataArray (dim_0: 50, dim_1: 512, dim_2:...mensions without coordinates: dim_0, dim_1, dim_2, key=(<class 'numpy.ndarray'> (262144,) int64, <class 'numpy.ndarray'> (262144,) int64, <class 'numpy.ndarray'> (262144,) int64))
    743         else:
    744             # xarray-style array indexing
--> 745             return self.isel(indexers=self._item_key_to_dict(key))
        self.isel = <bound method DataArray.isel of <xarray.DataArray (dim_0: 50, dim_1: 512, dim_2: 512)>
<class 'numpy.ndarray'> (50, 512, 512) uint16
Dimensions without coordinates: dim_0, dim_1, dim_2>
        global indexers = undefined
        self._item_key_to_dict = <bound method DataArray._item_key_to_dict of <xarray.DataArray (dim_0: 50, dim_1: 512, dim_2: 512)>
<class 'numpy.ndarray'> (50, 512, 512) uint16
Dimensions without coordinates: dim_0, dim_1, dim_2>
        key = (<class 'numpy.ndarray'> (262144,) int64, <class 'numpy.ndarray'> (262144,) int64, <class 'numpy.ndarray'> (262144,) int64)
    746 
    747     def __setitem__(self, key: Any, value: Any) -> None:

~/mambaforge/envs/micro/lib/python3.9/site-packages/xarray/core/dataarray.py in isel(self=<xarray.DataArray (dim_0: 50, dim_1: 512, dim_2:...mensions without coordinates: dim_0, dim_1, dim_2, indexers={'dim_0': <class 'numpy.ndarray'> (262144,) int64, 'dim_1': <class 'numpy.ndarray'> (262144,) int64, 'dim_2': <class 'numpy.ndarray'> (262144,) int64}, drop=False, missing_dims='raise', **indexers_kwargs={})
   1206         # lists, or zero or one-dimensional np.ndarray's
   1207 
-> 1208         variable = self._variable.isel(indexers, missing_dims=missing_dims)
        variable = undefined
        self._variable.isel = <bound method Variable.isel of <xarray.Variable (dim_0: 50, dim_1: 512, dim_2: 512)>
<class 'numpy.ndarray'> (50, 512, 512) uint16>
        indexers = {'dim_0': <class 'numpy.ndarray'> (262144,) int64, 'dim_1': <class 'numpy.ndarray'> (262144,) int64, 'dim_2': <class 'numpy.ndarray'> (262144,) int64}
        missing_dims = 'raise'
   1209 
   1210         coords = {}

~/mambaforge/envs/micro/lib/python3.9/site-packages/xarray/core/variable.py in isel(self=<xarray.Variable (dim_0: 50, dim_1: 512, dim_2: 512)>
<class 'numpy.ndarray'> (50, 512, 512) uint16, indexers={'dim_0': <class 'numpy.ndarray'> (262144,) int64, 'dim_1': <class 'numpy.ndarray'> (262144,) int64, 'dim_2': <class 'numpy.ndarray'> (262144,) int64}, missing_dims='raise', **indexers_kwargs={})
   1191 
   1192         key = tuple(indexers.get(dim, slice(None)) for dim in self.dims)
-> 1193         return self[key]
        self = <xarray.Variable (dim_0: 50, dim_1: 512, dim_2: 512)>
<class 'numpy.ndarray'> (50, 512, 512) uint16
        key = (<class 'numpy.ndarray'> (262144,) int64, <class 'numpy.ndarray'> (262144,) int64, <class 'numpy.ndarray'> (262144,) int64)
   1194 
   1195     def squeeze(self, dim=None):

~/mambaforge/envs/micro/lib/python3.9/site-packages/xarray/core/variable.py in __getitem__(self=<xarray.Variable (dim_0: 50, dim_1: 512, dim_2: 512)>
<class 'numpy.ndarray'> (50, 512, 512) uint16, key=(<class 'numpy.ndarray'> (262144,) int64, <class 'numpy.ndarray'> (262144,) int64, <class 'numpy.ndarray'> (262144,) int64))
    784         """
    785         dims, indexer, new_order = self._broadcast_indexes(key)
--> 786         data = as_indexable(self._data)[indexer]
        data = undefined
        global as_indexable = <function as_indexable at 0x7fd431b00550>
        self._data = <class 'numpy.ndarray'> (50, 512, 512) uint16
        indexer = OuterIndexer((<class 'numpy.ndarray'> (262144,) int64, <class 'numpy.ndarray'> (262144,) int64, <class 'numpy.ndarray'> (262144,) int64))
    787         if new_order:
    788             data = np.moveaxis(data, range(len(new_order)), new_order)

~/mambaforge/envs/micro/lib/python3.9/site-packages/xarray/core/indexing.py in __getitem__(self=NumpyIndexingAdapter(array=<class 'numpy.ndarray'> (50, 512, 512) uint16), key=(<class 'numpy.ndarray'> (262144, 1, 1) int64, <class 'numpy.ndarray'> (1, 262144, 1) int64, <class 'numpy.ndarray'> (1, 1, 262144) int64))
   1158     def __getitem__(self, key):
   1159         array, key = self._indexing_array_and_key(key)
-> 1160         return array[key]
        array = <class 'numpy.ndarray'> (50, 512, 512) uint16
        key = (<class 'numpy.ndarray'> (262144, 1, 1) int64, <class 'numpy.ndarray'> (1, 262144, 1) int64, <class 'numpy.ndarray'> (1, 1, 262144) int64)
   1161 
   1162     def __setitem__(self, key, value):

MemoryError: Unable to allocate 32.0 PiB for an array with shape (262144, 262144, 262144) and data type uint16
```

**Comment by andy-sweet** [comment]
The cause of the last reported error is due to how xarray handles indices of type `tuple[ndarray, ndarray, ndarray]` differently to numpy (orthogonal vs. vectorized). See the similar napari bug #2190 or the [xarray docs for more details](https://numpy.org/doc/stable/reference/generated/numpy.ix_.html).

The obvious/simple fix here is to use the same approach as in #2191 - special case the vectorized indices when data is an `xarray.DataArray`. Using it for `paint` but not for `fill` seems inconsistent, so my suggestion is to do that for now, unless there's a good reason not to. As reviewers of that PR, @sofroniewn, @jni : any thoughts on this?

I guess the dream was to support a more general fix like in #2376, but that got closed, so we may also want to open an issue to track that kind of approach.

**Comment by jni** [comment]
Oh hello!

Yup, might as well unify with #2191. I also agree with your idea of raising an issue from #2376 to track better approaches.
